lib/Doctrine/ORM/UnitOfWork.php

```diff
@@ -1569,6 +1569,14 @@ public function addToIdentityMap($entity)
         $className     = $classMetadata->rootEntityName;
 
         if (isset($this->identityMap[$className][$idHash])) {
+            if ($this->identityMap[$className][$idHash] !== $entity) {
+                throw new RuntimeException(sprintf(
```

Comment:
reposting part of my comment from #10797: Maybe we could even try to be smarter here and allow a non-proxy to replace a proxy? Because it is really easy to put an invalid into the `$identityMap` with `getReference`, and we could try to heal this inconsistency here

Reply:
I don’t think that’s possible: 

Even if an object started out as an uninitialized proxy, it may turn into a fully initialized instance and is the only object instance representing that particular entity identity. It may have gone into different associations, collections and such – we cannot swap all those out for another object.

Reply:
whether or not the Proxy is initialized can be checked with `__isInitialized()`, so that could be solved, but yeah, it would probably get quite involved and in the end it would only be a workaround for the fact that `getReference` does not give any guarantees that the referenced entity actually exists.